List only canonical timezones in timezone dropdowns

jonatanschroeder opened the pull request:
The institution/course/course instance timezone is used in different parts of the code, sometimes in conversions done via sprocs/queries (i.e., using Postgres), sometimes via the formatter package (i.e., using JS Intl). Postgres supports some timezones that are not supported by Intl. If a user selects one of these timezones, this can cause an issue in pages that use the formatter package (which will gradually be used more often, so this may happen more often for these timezones).

This PR changes the lib function that populates dropdowns for timezones to filter only those timezones supported by both Postgres and Intl.

AFAIK there is no restriction in sync currently that checks if the timezone is valid, and this PR does not attempt to add this restriction in sync. So this PR attempts to avoid the issue when the UI is used to update the timezone, but not for JSON updates. It also does not address server-side validation of UI values.

I did a bit of checking locally to check which timezones are not supported, and other than a timezone named "Factory", all other unsupported timezones have the `posix/` prefix.

I should point out that the production server (at least the CA one) does not currently have the `posix/*` or `Factory` timezones, I'm not sure why. What this tells me is that this is unlikely to be a problem in production at this point.

Fixes #11619.

jonatanschroeder commented on apps/prairielearn/src/lib/timezones.ts:
```diff
@@ -20,7 +20,23 @@ async function getAvailableTimezonesFromDB(): Promise<Timezone[]> {
-    memoizedAvailableTimezones = await getAvailableTimezonesFromDB();
+    // Different portions of the code use the timezone in either PostgreSQL or
+    // Intl, so we return only those timezones that are supported by both. While
+    // Intl provides a list of supported timezones via Intl.supportedValuesOf(),
+    // the list is not exhaustive and may not include some timezones supported
+    // by JS as aliases. Instead, we attempt to create a new DateTimeFormat
```
To clarify about this limitation: a first version of this change filtered by `supportedValuesOf()`, which caused a large list of timezones that [are aliases to other timezones](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal/ZonedDateTime#time_zones_and_offsets) not to be included. For example, some timezones like `MST` or `US/Arizona` are actually aliases to `America/Phoenix`, and are supported by `Intl.DateTimeFormat()`, but are not listed in `supportedValuesOf()`. I believe that it makes sense to give instructors access to these aliases, though I see the value of simplifying the dropdown list to show a shorter (but still comprehensive) list. I'd welcome any thoughts.

mwest1066 replied:
I do like the idea of having a shorter list if possible -- it's very long at the moment. But I haven't had a chance to look at the `supportedValuesOf()` list to see whether it's missing important timezones.

mwest1066 replied:
Oh cool. Looking through that list I think it would be great to only use the Canonical names. This both keeps the list shorter and removes some of the more confusing things (like "EST" which is a link to "America/Panama", which is not what anyone would want). I think the non-canonical names are mainly there for backward compatibility, which we don't care about in this context.

jonatanschroeder replied:
Any chance of checking what users are using, to double check that nobody is using a linked timezone? This wouldn't be a problem in most cases, but if the dropdown does not list the timezone the have currently on it may cause confusion in the Settings page.

```sql
SELECT DISTINCT display_timezone FROM pl_courses UNION
SELECT DISTINCT display_timezone FROM course_instances UNION
SELECT DISTINCT display_timezone FROM institutions;
```

jonatanschroeder replied:
One option is to have two lists of timezones: accepted timezones (all including links) and listed timezones (only canonical). The dropdown could list only canonical entries, but if the current value for the course/course instance is not canonical it could include the currently set timezone with appropriate values. In other words, as long as the timezone the current course has is valid, the dropdown would list it in conjunction with canonical timezones, but as soon as the instructor changes it to a canonical entry, it would no longer be listed.

jonatanschroeder replied:
Done in 7c5300d3cab2e37c49ca1cdb800ef32c3b05d25b.

nwalters512 replied:
I like your solution!

Just because I was curious, I ran that query anyways on `us.prairielearn.com`. Out of 35 distinct time zones, 14 are non-canonical.

```
America/Montreal
Asia/Harbin
Canada/Pacific
Europe/Amsterdam
Europe/Stockholm
Hongkong
Portugal
US/Arizona
US/Central
US/Eastern
US/Michigan
US/Mountain
US/Pacific
UTC
```

mwest1066 replied:
Yep, great solution!

jonatanschroeder commented:
> I should point out that the production server (at least the CA one) does not currently have the `posix/*` or `Factory` timezones, I'm not sure why. What this tells me is that this is unlikely to be a problem in production at this point.

Actually, the "Factory" timezone is available, and causes a similar error in production (I just tested with [a course that is no longer active](https://ca.prairielearn.com/pl/course/8/course_admin/settings)).

mwest1066 commented:
> Actually, the "Factory" timezone is available, and causes a similar error in production (I just tested with [a course that is no longer active](https://ca.prairielearn.com/pl/course/8/course_admin/settings)).

I suspected you were testing this 😂

<img width="664" alt="image" src="https://github.com/user-attachments/assets/31c9c3f1-9117-4f89-ab30-18434b1c7d39" />

nwalters512 commented on apps/prairielearn/src/lib/timezones.ts:
```diff
@@ -18,13 +18,36 @@ async function getAvailableTimezonesFromDB(): Promise<Timezone[]> {
+ */
+export async function getCanonicalTimezones(alwaysInclude?: string[]): Promise<Timezone[]> {
+  const availableTimezones = await getAvailableTimezones();
+  const canonicalTimezones = Intl.supportedValuesOf('timeZone');
```
This returns neither `Etc/UTC` (canonical) or `UTC` (link), but I tested both those locally and they still work with `Intl.DateTimeFormat`. This makes me nervous that there are others that are for some reason missing from `supportedValuesOf`.

nwalters512 replied:
Turns out V8 is just being weird:

- https://github.com/tc39/ecma402/issues/778
- https://issues.chromium.org/issues/42203023

Seems like only Firefox has the expected behavior right now. I'd be inclined to manually include `UTC` in the list; note that in JS, the other variants of UTC are canonicalized to `UTC` "for historical reasons" (https://tc39.es/ecma402/#sec-use-of-iana-time-zone-database).

reteps commented:
Does this also need to update this comment to be consistent with the JSON Schema? https://github.com/PrairieLearn/PrairieLearn/blob/6d0c67ff77194c772d87312165479d51e1d33c13/apps/prairielearn/src/schemas/schemas/infoCourse.json#L27

mwest1066 commented:
> Even though you said there is no restriction, is it worth pointing people at a list that is consistent with the frontend?

I think we just need to say that it's anything "Canonical" in the Wikipedia list?

jonatanschroeder commented:
> Does this also need to update this comment to be consistent with the JSON Schema?

I updated the description in ea3d23098d76eeafcd611daf9b170d0a387159d3. I listed canonical as preferred, since we still support non-canonical. I also removed the listed default, since it's no longer accurate.
